pinot-query-runtime/src/main/java/org/apache/pinot/query/service/QueryDispatcher.java

```diff
@@ -220,23 +252,94 @@ public static MailboxReceiveOperator createReduceStageOperator(MailboxService<Tr
 
   public void shutdown() {
     for (DispatchClient dispatchClient : _dispatchClientMap.values()) {
-      dispatchClient._managedChannel.shutdown();
+      dispatchClient._channel.shutdown();
     }
     _dispatchClientMap.clear();
   }
 
+  @VisibleForTesting
+  DispatchClient getOrCreateDispatchClient(String host, int port) {
+    String key = String.format("%s_%d", host, port);
+    return _dispatchClientMap.computeIfAbsent(key, k -> new DispatchClient(host, port));
+  }
+
   public static class DispatchClient {
-    private final PinotQueryWorkerGrpc.PinotQueryWorkerBlockingStub _blockingStub;
-    private final ManagedChannel _managedChannel;
+    private ManagedChannel _channel;
+    private PinotQueryWorkerGrpc.PinotQueryWorkerStub _dispatchStub;
 
     public DispatchClient(String host, int port) {
-      ManagedChannelBuilder managedChannelBuilder = ManagedChannelBuilder.forAddress(host, port).usePlaintext();
-      _managedChannel = managedChannelBuilder.build();
-      _blockingStub = PinotQueryWorkerGrpc.newBlockingStub(_managedChannel);
+      _channel = ManagedChannelBuilder.forAddress(host, port).usePlaintext().build();
```

Comment:
nit: let's move the DispatchClient and DispatchObserver to their own file under `org.apache.pinot.query.service.dispatch`. makes the code cleaner.